File: 0002-add-two-numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode temp = dummy;
            int carry = 0;
        while(l1 != null || l2 != null || carry > 0) {
            int sum = 0;
            if(l1 != null) {
                sum += l1.val;//3
                l1 = l1.next;
            }

            if(l2 != null) {
                sum += l2.val;//4
                l2 = l2.next;
            }
            sum += carry;//8
            int dig = sum%10;//0
            carry = sum/10;//1
            temp.next = new ListNode(dig);//-1 -> 7 -> 0
            temp = temp.next;
        }
        return dummy.next;
    }
}

File: 0002-add-two-numbers/solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        let dummy = ListNode(-1)
        var temp: ListNode? = dummy

        var p1 = l1
        var p2 = l2
        var carry = 0
        // why (OR) -> sometimes list1 elements can be more, sometimes list2 will have more
        // if carry > 0 is not included 9+9 = 18 will miss carry here
        while(p1 != nil || p2 != nil || carry > 0) {
            var sum = 0
            if let node1 = p1 {
                sum += node1.val
                p1 = node1.next
            }

            if let node2 = p2 {
                sum += node2.val
                p2 = node2.next
            }

            // calculate sum with old carry
            sum += carry
            // get a new digit
            var digit = sum%10
            // get new carry
            carry = sum/10
            temp?.next = ListNode(digit)
            temp = temp?.next
        }
        return dummy.next
    }
}
